text/0000-xor-urls/0000-xor-urls.md

```diff
@@ -0,0 +1,115 @@
+
+## Detailed design
+
+### XOR address encoding
+The XOR address shall be encoded in the XOR-URL, [z-base32 encoding](http://philzimmermann.com/docs/human-oriented-base-32-encoding.txt) seems to be a good choice as it is case insensitive, as opposed to other case-sensitive encodings like base58btc or base64url, and it was designed to be easier for human use permuting the alphabet so that the easier characters are the ones that occur more frequently.
+
+### ImmutableData XOR-URL
+XOR-URLs for ImmutableData’s are the simplest ones since they don’t need any additional information to uniquely identify them on the network, as opposed to MutableData’s that also have a type tag. Therefore an ImmutableData XOR-URL can be simply defined as `safe://<encoded ImmutableData XOR addr>`.
+
+### MutableData XOR-URL
+As already mentioned above, a string based on the XOR address along with a type tag is needed to uniquely identify a MutableData on the network, therefore the XOR-URL for a MD needs to include the type tag in it.
+
+When a MutableData is fetched, if it’s not an NFS/Files container with an `index.html` file, the resolver (e.g. the `webFetch` function) can return the MutableData’s raw data so the browser (or any client application) can render it in a specific way (see below for more details of what’s being proposed here in this regard).
+
+### MutableData versions
+MutableData’s are versioned and therefore this shall be also accountable in the XOR-URL format to allow any MD URL to (optionally) reference a specific version. This will be also necessary in the future when append-only type of data is made available on the network.
+
+The version value can be used to enforce a specific version to be retrieved, and otherwise fail if that specific version is not found. On the other hand, the latest version will be retrieved if the version value is omitted from the XOR-URL.
+
+### Paths
+Given that a referenced MutableData can effectively be the root NFS/Files container of the files of a safesite, any MD XOR-URL can also specify a path which needs to be resolved, just like when using the DNS/public-names system to resolve the path in a URL.
+
+### Browsable content
+Currently, when a public-name URL is resolved to an NFS/Files container which doesn’t have an `index.html` file, the browser simply shows an error stating that the safesite content was not found.
+
+For a MD XOR-URL that doesn’t have a path and there is no `index.html` file, the resolver can return the raw content of the MD (i.e. its key-value entries), and the browser can automatically generate an HTML page which makes the content browsable, generating links to other data when an entry’s key/value is a `safe://` string, in an analogous/similar way to how web servers on the current internet allow browsing on folders/directories.
```

Comment:
> For a MD XOR-URL that doesn’t have a path and there is no `index.html` file

To what extent should this reference the other RFC? As we could talk about `default` entries vs the specific `index.html`....

Reply:
Do you mean a link to the PNS RFC PR? I can do that yes, although as far as this RFC is concern I guess it's just if there was no index.html or default resolved from the NFS container found, right? or you are suggesting to have some more details here about the resolution too?

Reply:
> it's just if there was no index.html or default resolved from the NFS container found, right?

right.

> or you are suggesting to have some more details here about the resolution too?

Nono, just aligning where makes sense 👍